File: third-party/thrift/src/thrift/lib/hack/src/transport/TFramedTransport.php
<?hh
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

// @oss-enable: use namespace FlibSL\{C, Math, Str, Vec};

/**
 * Framed transport. Writes and reads data in chunks that are stamped with
 * their length.
 *
 * @package thrift.transport
 */
<<Oncalls('thrift')>> // @oss-disable
class TFramedTransport
  extends TTransport
  implements TTransportStatus, IThriftBufferedTransport {

  /**
   * Underlying transport object.
   *
   * @var TTransport
   */
  protected TTransport $transport_;

  /**
   * Buffer for read data.
   *
   * @var string
   */
  protected string $rBuf_ = '';

  /**
   * Position in rBuf_ to read the next char
   *
   * @var int
   */
  protected int $rIndex_ = 0;

  /**
   * Buffer for queued output data
   *
   * @var string
   */
  protected string $wBuf_ = '';

  /**
   * Whether to frame reads
   *
   * @var bool
   */
  private bool $read_;

  /**
   * Whether to frame writes
   *
   * @var bool
   */
  private bool $write_;

  /**
   * Constructor.
   *
   * @param TTransport $transport Underlying transport
   */
  public function __construct(
    ?TTransport $transport = null,
    bool $read = true,
    bool $write = true,
  )[] {
    $this->transport_ = $transport ?: new TNullTransport();
    $this->read_ = $read;
    $this->write_ = $write;
  }

  <<__Override>>
  public function isOpen()[]: bool {
    return $this->transport_->isOpen();
  }

  <<__Override>>
  public function open()[zoned_shallow]: void {
    $this->transport_->open();
  }

  <<__Override>>
  public function close()[zoned_shallow]: void {
    $this->transport_->close();
  }

  public function isReadable()[leak_safe]: bool {
    if (Str\length($this->rBuf_) > 0) {
      return true;
    }

    if ($this->transport_ is TTransportStatus) {
      return $this->transport_->isReadable();
    }

    return true;
  }

  public function isWritable()[leak_safe]: bool {
    if ($this->transport_ is TTransportStatus) {
      return $this->transport_->isWritable();
    }

    return true;
  }

  /**
   * Reads from the buffer. When more data is required reads another entire
   * chunk and serves future reads out of that.
   *
   * @param int $len How much data
   */
  <<__Override>>
  public function read(int $len)[zoned_shallow]: string {
    if (!$this->read_) {
      return $this->transport_->read($len);
    }

    if (Str\length($this->rBuf_) === 0) {
      $this->readFrame();
    }

    // Return substr
    $out = PHP\substr($this->rBuf_, $this->rIndex_, $len);
    $this->rIndex_ += $len;

    if (Str\length($this->rBuf_) <= $this->rIndex_) {
      $this->rBuf_ = '';
      $this->rIndex_ = 0;
    }
    return $out;
  }

  /**
   * Peek some bytes in the frame without removing the bytes from the buffer
   *
   * @param int $len   length to peek
   * @param int $start the start position of the returned string
   */
  public function peek(int $len, int $start = 0)[zoned_shallow]: string {
    if (!$this->read_) {
      return '';
    }

    if (Str\length($this->rBuf_) === 0) {
      $this->readFrame();
    }

    // Return substr
    $out = PHP\substr($this->rBuf_, $this->rIndex_ + $start, $len);

    return $out;
  }

  /**
   * Put previously read data back into the buffer
   *
   * @param string $data data to return
   */
  public function putBack(string $data)[write_props]: void {
    if (Str\length($this->rBuf_) === 0) {
      $this->rBuf_ = $data;
    } else {
      $this->rBuf_ = ($data.(string)PHP\substr($this->rBuf_, $this->rIndex_));
    }
    $this->rIndex_ = 0;
  }

  /**
   * Reads a chunk of data into the internal read buffer.
   */
  private function readFrame()[zoned_shallow]: void {
    $buf = $this->transport_->readAll(4);
    $val = PHP\unpack('N', $buf);
    $sz = $val[1];

    $this->rBuf_ = $this->transport_->readAll($sz);
    $this->rIndex_ = 0;
  }

  /**
   * Writes some data to the pending output buffer.
   *
   * @param string $buf The data
   * @param int    $len Limit of bytes to write
   */
  <<__Override>>
  public function write(string $buf, ?int $len = null)[zoned_shallow]: void {
    if (!$this->write_) {
      $this->transport_->write($buf);
      return;
    }

    if ($len !== null && $len < Str\length($buf)) {
      $buf = PHP\substr($buf, 0, $len);
    }
    $this->wBuf_ .= (string)$buf;
  }

  /**
   * Writes the output buffer to the stream in the format of a 4-byte length
   * followed by the actual data.
   */
  <<__Override>>
  public function flush()[zoned_shallow]: void {
    if (!$this->write_ || Str\length($this->wBuf_) === 0) {
      $this->transport_->flush();
      return;
    }

    $out = (string)PHP\pack('N', Str\length($this->wBuf_));
    $out .= $this->wBuf_;

    // Note that we clear the internal wBuf_ prior to the underlying write
    // to ensure we're in a sane state (i.e. internal buffer cleaned)
    // if the underlying write throws up an exception
    $this->wBuf_ = '';
    $this->transport_->write($out);
    $this->transport_->flush();
  }

}

File: third-party/thrift/src/thrift/lib/javadeprecated/src/main/java/com/facebook/thrift/transport/TFramedTransport.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.thrift.transport;

import com.facebook.thrift.TByteArrayOutputStream;

/**
 * TFramedTransport is a buffered TTransport that ensures a fully read message every time by
 * prefixing messages with a 4-byte frame size.
 */
public class TFramedTransport extends TTransport {

  protected static final int DEFAULT_MAX_LENGTH = 0x7FFFFFFF;

  private int maxLength_;

  /** Underlying transport */
  protected TTransport transport_ = null;

  /** Buffer for output */
  protected final TByteArrayOutputStream writeBuffer_ = new TByteArrayOutputStream(1024);

  /** Buffer for input */
  protected TMemoryInputTransport readBuffer_ = new TMemoryInputTransport(new byte[0]);

  public static class Factory extends TTransportFactory {
    private int maxLength_;

    public Factory() {
      maxLength_ = TFramedTransport.DEFAULT_MAX_LENGTH;
    }

    public Factory(int maxLength) {
      maxLength_ = maxLength;
    }

    @Override
    public TTransport getTransport(TTransport base) {
      return new TFramedTransport(base, maxLength_);
    }
  }

  /** Constructor wraps around another tranpsort */
  public TFramedTransport(TTransport transport, int maxLength) {
    transport_ = transport;
    maxLength_ = maxLength;
  }

  public TFramedTransport(TTransport transport) {
    transport_ = transport;
    maxLength_ = TFramedTransport.DEFAULT_MAX_LENGTH;
  }

  @Override
  public void open() throws TTransportException {
    transport_.open();
  }

  @Override
  public boolean isOpen() {
    return transport_.isOpen();
  }

  @Override
  public void close() {
    transport_.close();
  }

  @Override
  public int read(byte[] buf, int off, int len) throws TTransportException {
    int got = readBuffer_.read(buf, off, len);
    if (got > 0) {
      return got;
    }

    // Read another frame of data
    readFrame();

    return readBuffer_.read(buf, off, len);
  }

  @Override
  public byte[] getBuffer() {
    return readBuffer_.getBuffer();
  }

  @Override
  public int getBufferPosition() {
    return readBuffer_.getBufferPosition();
  }

  @Override
  public int getBytesRemainingInBuffer() {
    return readBuffer_.getBytesRemainingInBuffer();
  }

  @Override
  public void consumeBuffer(int len) {
    readBuffer_.consumeBuffer(len);
  }

  private final byte[] i32buf = new byte[4];

  protected void readFrame() throws TTransportException {
    transport_.readAll(i32buf, 0, 4);
    int size = decodeWord(i32buf);

    if (size < 0) {
      throw new TTransportException(String.format("Read a negative frame size (%d)!", size));
    }

    if (size > maxLength_) {
      throw new TTransportException(
          String.format("Frame size (%d) larger than max length (%d)!", size, maxLength_));
    }

    byte[] buff = new byte[size];
    transport_.readAll(buff, 0, size);
    readBuffer_.reset(buff);
  }

  @Override
  public void write(byte[] buf, int off, int len) throws TTransportException {
    writeBuffer_.write(buf, off, len);
  }

  @Override
  public void flush() throws TTransportException {
    byte[] buf = writeBuffer_.get();
    int len = writeBuffer_.len();
    writeBuffer_.reset();

    encodeWord(len, i32buf);
    transport_.write(i32buf, 0, 4);
    transport_.write(buf, 0, len);
    transport_.flush();
  }

  /** Functions to encode/decode int32 and int16 to/from network byte order */
  public static final void encodeWord(final int frameSize, final byte[] buf) {
    buf[0] = (byte) (0xff & (frameSize >> 24));
    buf[1] = (byte) (0xff & (frameSize >> 16));
    buf[2] = (byte) (0xff & (frameSize >> 8));
    buf[3] = (byte) (0xff & (frameSize));
  }

  public static final int decodeWord(final byte[] buf) {
    return decodeWord(buf, 0);
  }

  public static final int decodeWord(final byte[] buf, int off) {
    return ((buf[0 + off] & 0xff) << 24)
        | ((buf[1 + off] & 0xff) << 16)
        | ((buf[2 + off] & 0xff) << 8)
        | ((buf[3 + off] & 0xff));
  }

  public static final short decodeShort(final byte[] buf) {
    return decodeShort(buf, 0);
  }

  public static final short decodeShort(final byte[] buf, int off) {
    return (short) (((buf[0 + off] & 0xff) << 8) | ((buf[1 + off] & 0xff)));
  }

  public static final void encodeShort(final int value, final byte[] buf) {
    buf[0] = (byte) (0xff & (value >> 8));
    buf[1] = (byte) (0xff & (value));
  }
}
